server/loginserver.ts

```diff
@@ -38,69 +36,74 @@ function parseJSON(/** @type {string} */ json) {
 	return data;
 }
 
-/** @typedef {[AnyObject?, number, Error?]} LoginServerResponse */
+type LoginServerResponse = [AnyObject | null, number, Error | null];
+
+interface IncomingMessage extends NodeJS.ReadableStream {
+	statusCode: number;
+}
 
 class LoginServerInstance {
+	uri: string;
+	requestQueue: [AnyObject, (val: LoginServerResponse) => void][];
+	requestTimer: NodeJS.Timer | null;
+	requestLog: string;
+	lastRequest: number;
+	openRequests: number;
+	disabled: false;
+
 	constructor() {
 		this.uri = Config.loginserver;
-		/**
-		 * @type {[AnyObject, (val: LoginServerResponse) => void][]}
-		 */
 		this.requestQueue = [];
-
 		this.requestTimer = null;
-		/** @type {string} */
 		this.requestLog = '';
 		this.lastRequest = 0;
 		this.openRequests = 0;
 		this.disabled = false;
 	}
 
-	/**
-	 * @param {string} action
-	 * @param {AnyObject?} data
-	 * @return {Promise<LoginServerResponse>}
-	 */
-	instantRequest(action, data = null) {
+	instantRequest(action: string, data: AnyObject | null = null): Promise<LoginServerResponse> {
 		if (this.openRequests > 5) {
-			return Promise.resolve(/** @type {LoginServerResponse} */ (
+			return Promise.resolve((
 				[null, 0, new RangeError("Request overflow")]
 			));
 		}
 		this.openRequests++;
 		let dataString = '';
 		if (data) {
-			for (let i in data) {
+			for (const i in data) {
 				dataString += '&' + i + '=' + encodeURIComponent('' + data[i]);
 			}
 		}
-		const urlObject = url.parse(this.uri + 'action.php?act=' + action + '&serverid=' + Config.serverid + '&servertoken=' + encodeURIComponent(Config.servertoken) + '&nocache=' + new Date().getTime() + dataString);
+
+		const urlString = this.uri + 'action.php?' +
+		'?act=action&serverid=' + Config.serverid +
```

Comment:
I think this was indented before?

Reply:
url is reserved for the module, so used `urlObject`.

Reply:
Ah, right. didn't catch that. Maybe `actionUrl`, because its for the action.php endpoint?

Reply:
Don't have any strong opinions on it, so sure.